framework/src/base/Assembly.C

```diff
@@ -1503,6 +1557,23 @@ Assembly::addJacobianBlock(SparseMatrix<Number> & jacobian, unsigned int ivar, u
   }
   else
     jacobian.add_matrix(ke, di);
+
+  DenseMatrix<Number> & keg = jacobianBlockNonlocal(ivar, jvar);
+
+  std::vector<dof_id_type> dg;
+  for (unsigned int i=0; i<_dof_map.n_dofs(); i++)
+    dg.push_back(i);
+
+  dof_map.constrain_element_matrix(keg, di, dg, false);
+
+  if (scaling_factor != 1.0)
+  {
+    _tmp_Ke = keg;
+    _tmp_Ke *= scaling_factor;
+    jacobian.add_matrix(_tmp_Ke, di, dg);
+  }
+  else
+    jacobian.add_matrix(keg, di, dg);
```

Comment:
The above new code in the `addJacobianBlock` method should only operate when you have Nonlocal objects, right? I believe this is the cause of a lot of your time out errors based on my testing. Maybe insert a check.